libraries/AP_Motors/AP_MotorsMatrix.cpp

```diff
@@ -1083,6 +1083,50 @@ bool AP_MotorsMatrix::setup_octaquad_matrix(motor_frame_type frame_type)
         add_motors(motors, ARRAY_SIZE(motors));
         break;
     }
+    case MOTOR_FRAME_TYPE_X_COR: {
+        _frame_type_string = "X_COR";
+        static const AP_MotorsMatrix::MotorDef motors[] {
+            {   45, AP_MOTORS_MATRIX_YAW_FACTOR_CCW, 1 },
+            {  -45, AP_MOTORS_MATRIX_YAW_FACTOR_CW,  7 },
+            { -135, AP_MOTORS_MATRIX_YAW_FACTOR_CCW, 5 },
+            {  135, AP_MOTORS_MATRIX_YAW_FACTOR_CW,  3 },
+            {  -45, AP_MOTORS_MATRIX_YAW_FACTOR_CW,  8 },
+            {   45, AP_MOTORS_MATRIX_YAW_FACTOR_CCW, 2 },
+            {  135, AP_MOTORS_MATRIX_YAW_FACTOR_CW,  4 },
+            { -135, AP_MOTORS_MATRIX_YAW_FACTOR_CCW, 6 },
+        };
+        add_motors(motors, ARRAY_SIZE(motors));
+        // scale top layer to prevent a beat frequency between layers in co-rotating setups
+        for (uint8_t i=0; i<4; i++) {
+            _roll_factor[i] *= AP_MOTORS_FRAME_OCTAQUAD_COROTATING_SCALE_FACTOR;
+            _pitch_factor[i] *= AP_MOTORS_FRAME_OCTAQUAD_COROTATING_SCALE_FACTOR;
+            _yaw_factor[i] *= AP_MOTORS_FRAME_OCTAQUAD_COROTATING_SCALE_FACTOR;
+            _throttle_factor[i] *= AP_MOTORS_FRAME_OCTAQUAD_COROTATING_SCALE_FACTOR;
```

Comment:
I'm slightly reluctant to have throttle factors in the c++ without mixer updates. Currently we do the mixing and critically the saturation handling first. Then we apply the throttle factors afterwards. This means we get the saturation cases wrong, limiting too soon at high throttle and too late at low throttle. It would need some re-work in the mixer, but we should be able to include the throttle factor in the saturation calculations. 

I'm also a little surprised you have to scale roll/pitch/yaw too. I would have thought just throttle would be enough to avoid a beat frequency between motors on the same arm. With the same roll/pitch/yaw factors both motors on each arm would change the same amount, so they should always have the same split due to the throttle factor.

Reply:
just doing the throttle factor didn't work well in testng. I originally did the lua script with just the throttle factor for the same reasons you suggested, I then tried a wide variety of MOT_X8C_TOPMUL and MOT_X8C_BOTMUL on a real X8. The most extreme case was setting MOT_X8C_BOTMUL to 0 to kill the bottom layer completely, and it flew extremely badly when I didn't scale the roll/pitch/yaw factors
Another case is when the throttle output for a corner of the vehicle goes to zero. This is most easily produced in ACRO mode, but can happen in any mode with extreme manoeuvrers. In that case if we only scale the throttle factor then we'll end up with both motors on a corner running at the same output, and will get the beat frequency.